clis/teliod/src/main.rs

```diff
@@ -0,0 +1,183 @@
+
+use anyhow::Result;
+use clap::Parser;
+use std::{
+    fs::{self, File},
+    str::FromStr,
+    time::Duration,
+};
+use tracing::{error, info, level_filters::LevelFilter, warn};
+
+use daemon_common::{
+    comms::DaemonSocket,
+    daemon::{Daemon, ProcessType},
+};
+
+use telio::{device::Device, telio_model::features::Features};
+
+use serde::{de, Deserialize, Deserializer, Serialize};
+
+struct TeliodDaemon;
+
+impl Daemon for TeliodDaemon {
+    const NAME: &'static str = "teliod";
+}
+
+#[derive(Deserialize)]
+struct TeliodDaemonConfig {
+    #[serde(deserialize_with = "deserialize_log_level")]
+    log_level: LevelFilter,
+    log_file_path: String,
+}
+
+fn deserialize_log_level<'de, D>(deserializer: D) -> Result<LevelFilter, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    Deserialize::deserialize(deserializer).and_then(|s: String| {
+        LevelFilter::from_str(&s).map_err(|_| {
+            de::Error::unknown_variant(&s, &["error", "warn", "info", "debug", "trace", "off"])
+        })
+    })
+}
+
+/// Simple telio daemon for Linux and OpenWRT
+#[derive(Parser)]
+#[clap()]
+#[derive(Serialize, Deserialize)]
+enum Cmd {
+    #[clap(about = "Starts the daemon")]
+    Start { config_path: String },
+    #[clap(
+        about = "Forces daemon to add a log, added for testing to be, to be removed when the daemon will be more mature"
+    )]
+    HelloWorld { name: String },
+    #[clap(about = "Same as 'status' but in json string")]
+    Stop,
+}
+
+fn main() -> Result<()> {
+    match Cmd::parse() {
+        Cmd::Start { config_path } => {
+            if TeliodDaemon::is_running()? {
+                eprintln!("Teliod is already running, stop it by calling `teliod stop`");
+            } else {
+                let file = File::open(config_path)?;
+                let config: TeliodDaemonConfig = serde_json::from_reader(file)?;
+                start_daemon(config)?;
+            }
+        }
+        cmd => {
+            if TeliodDaemon::is_running()? {
+                let response = DaemonSocket::send_command(
+                    &TeliodDaemon::get_ipc_socket_path()?,
+                    &serde_json::to_string(&cmd)?,
+                )?;
+
+                if response.as_str() == "OK" {
+                    println!("Command executed successfully");
+                } else {
+                    eprintln!("Command failed");
+                }
+            } else {
+                eprintln!("Teliod daemon is not running");
+            }
+        }
+    };
+
+    Ok(())
+}
+
+/// Starts the Teliod daemon which will run actual Telio.
+///
+/// # Arguments
+///
+/// * `config` - Daemon config.
+fn start_daemon(config: TeliodDaemonConfig) -> Result<()> {
+    println!("Starting Teliod daemon...");
+    match TeliodDaemon::start() {
+        Ok(ProcessType::Api) => {
+            // Verifying that the server had been started, but we have no way of synchronizing
```

Comment:
I think this solution using time arrays can be improved. We can use a one shot channel or even a bool flag for synchronization.. otherwise a exponential backoff mechanism can also be used.

Reply:
Something like this maybe?